src/schema-utils.js

```diff
@@ -1,17 +1,82 @@
+const schemaDefaults = require('json-schema-defaults');
+const keys = require('lodash/keys');
+const filter = require('lodash/filter');
+const omitBy = require('lodash/omitBy');
+const merge = require('lodash/merge');
+const mapKeys = require('lodash/mapKeys');
+const Validator = require('ajv');
+const snakeCase = require('snake-case');
+const { ValidationError } = require('src/errors');
+
+
 function defaults(schema, d) {
-  // TODO
-  return d;
+  return merge({}, schemaDefaults(schema), d);
 }
 
 
 function omitReadOnly(schema, d) {
-  // TODO
-  return d;
+  // TODO support for values other than single-level object properties
+  return omitBy(d, (v, k) => propertyIsReadOnly(schema, k));
 }
 
 
 function validate(schema, d) {
-  // TODO
+  const errors = []
+    .concat(validationErrors(schema, d))
+    .concat(readOnlyErrors(schema, d));
+
+  if (errors.length) throw new ValidationError(errors);
+}
+
+
+function validationErrors(schema, d) {
+  const validator = new Validator({
+    allErrors: true,
+    jsonPointers: true
+  });
+
+  validator.validate(schema, d);
+  return (validator.errors || []).map(parseValidationError);
+}
+
+
+function readOnlyErrors(schema, d) {
+  // TODO support for values other than single-level object properties
+  return filter(keys(d), k => propertyIsReadOnly(schema, k))
+  .map(k => parseReadOnlyError({
+    name: k,
+    path: `/${k}`,
+    schemaPath: `#/${k}/readOnly`
+  }));
+}
+
+
+function parseReadOnlyError({name, path, schemaPath}) {
```

Comment:
I'm a bit confused by the name of this function -- it seems to be creating a read only error object rather than parsing one? Would `readOnlyError` perhaps be a better name?